Speaker Feedback: Add a daily cron job to notify speakers of new feedback

coreymckrill opened the pull request:
This creates a system that gathers feedback comments that have been approved but have not yet been included in a notification (using a `speaker_notified` comment meta value). From those comments it derives a list of the related session posts, and from that derives a list of speaker user accounts. Each speaker is then sent one email with a list of each of their sessions that has newly approved feedback on it. The cron job is set up to ensure this doesn't happen more than once per day. 6pm in the WordCamp site's local timezone was chosen for the send time so that organizers may have time to approve some feedback comments the same day that the event occurred and still have notifications go out.

This also includes a user meta value, `sft_speaker_notifications_opt_out` that will skip sending an email notification if it is set for a user. However, it does not include any UI for toggling the value. I've left that for a separate ticket/PR since it will need some design work.

Fixes #422 

### How to test the changes in this Pull Request:

1. Set up the data:
    * Two speaker posts with usernames that exist on the network.
    * Three or more session posts. One for each of the speakers, and one with both of the speakers. The session times should be a day or two before the current time to ensure they fall within the window when notification emails will be sent out.
    * Add some feedback comments to each session, but leave them unapproved.
1. Open up MailCatcher so you can see when email notifications get sent out.
1. You can trigger the cron job manually with wp-cli. On the docker environment it will be something like
    `wp --allow-root --url=https://2014.seattle.wordcamp.org cron event run sft_notify_speakers_approved_feedback`
1. When you trigger it before approving any feedback, no emails should be sent out.
1. Approve a feedback comment for each session, and then trigger the cron job again. This time there should be emails in MailCatcher. Notice that there should be one email per speaker, not per session. The email should list each session that has new feedback.
1. If you then immediately trigger the cron job again, no new emails should be sent out, because the approved feedback comments have all been marked as `speaker_notified`. If you unapprove comments and then reapprove them, it should not reset the notified status.
1. Try changing the timestamp of the sessions to get outside the window when email notifications will be sent (earliest session + 3 months). You'll need to use wp-cli to delete the `sft_speaker_notification_start_time` transient in order to test this.

coreymckrill committed:
Add cron job to notify speakers of new feedback daily

coreymckrill committed:
Attempt to localize speaker notifications

coreymckrill committed:
Load cron file

coreymckrill committed:
Tweak email notification

coreymckrill committed:
Tweak speaker_notifications_are_enabled()

coreymckrill committed:
Abstract front end feedback URL generation to a separate function

This way it can be used for #440 as well.

coreymckrill committed:
Fix comment sorting to retain comment ID

wp_list_pluck was not correctly combining the comment ID and comment
post ID as key => value pairs. The keys were just numeric incremental
starting from 0. I'm not sure why this is happening, but maybe it's
because the comment IDs are numeric, so PHP ignores the given indexes
and just uses the default? Using array_combine works around the issue.

coreymckrill committed:
Consider singular/plural for subject line translation

coreymckrill committed:
Add check for user meta opt out value

coreymckrill committed:
Make the feedback link for speakers a login link

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/cron.php:
```diff
@@ -0,0 +1,203 @@
+	foreach ( $speaker_session_ids as $user_id => $session_ids ) {
+		$speaker = get_user_by( 'id', $user_id );
+		// Todo: Implement a UI for toggling this user meta value.
+		if ( true === $speaker->sft_speaker_notifications_opt_out ) {
```
This will opt the user out for all WordCamps in the future too, right? since it's a per-user setting

I _think_ that's fine, we'll just want to make note of that whatever UI lets users control this

ryelle replied:
Tried testing this with `wp user meta set 4 sft_speaker_notifications_opt_out true`, and it set the value as a string. I'm not sure if setting meta directly will preserve type, but you might want to switch this to `wp_validate_boolean( … )`

coreymckrill replied:
> This will opt the user out for all WordCamps in the future too, right?

Yeah, I believe this will be a value stored in the usermeta table that is shared across the entire WordPress/WordCamp network. So if the speaker opts out of notifications on one site, they will be opted out on other future WordCamp sites as well. But they'd also be able to opt back in from any site.

> you might want to switch this to wp_validate_boolean( … )

Good idea. ecbd74fd

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/cron.php:
```diff
@@ -0,0 +1,203 @@
+		}
+
+		$speaker_locale = get_user_locale( $speaker );
+		switch_to_locale( $speaker_locale );
```
💯

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/comment-meta.php:
```diff
@@ -143,6 +143,16 @@ function get_feedback_meta_field_schema( $context = 'all', $key = '' ) {
 			'required'          => false,
 		),
+		'speaker_notified' => array(
+			'sft_context'       => array( 'update' ),
```
This meta value is editable by the speaker via the API, like `helpful` — did you intend that? (I'm not sure if it matters, but it seemed unexpected)

coreymckrill replied:
I was aware of that when I added it, but it seems unlikely that it would ever be set through the API in a way that we don't intend. I guess we could add another value for `sft_context`. Maybe `internal` or something that indicates it's only used programmatically?

ryelle replied:
`internal` sounds good 👍

coreymckrill replied:
I guess another option would be to just not include it in the field schema, since, as mentioned, it's only for internal usage...

coreymckrill replied:
Decided to go with `internal`: 6c6207b5

ryelle reviewed (approved):
Works great! Have a few questions, but nothing blocking merge.

coreymckrill committed:
Sanitize boolean value before checking

coreymckrill committed:
Set context of `speaker_notified` meta to `internal`

This indicates that the field is only used programmatically, and
prevents it from being added or changed via an API update request.
